Set default values for Blaze Ad Preview when fetching AI suggestions fail

pmusolino opened the pull request:
Closes: #14616

## Description
In this PR, I updated the logic in the Blaze Campaign Creation form to improve the scenario where AI suggestions fail, preventing an empty and buggy preview from being displayed.

## Changes
- Updated `BlazeCampaignCreationFormViewModel` to enhance the AI suggestion logic by checking if the tagline, description, and ctaText are empty in case of failure, and then populating them with product information.
- Added a test `test_onLoad_does_not_fetch_AI_suggestions_again` to ensure AI suggestions are not fetched again if the tagline, description, and ctaText have already been populated.

## Testing information
1) Start the process to create a new Blaze campaign.
2) Before opening the product, enable the Network Link Conditioner or disable the internet connection.
3) Open the product.
4) Notice that after the alert, which mentions that it is unable to fetch AI data, the preview is populated with product information.
5) Try to continue by opening another view, then return to the Blaze Creation Form.
6) Notice that the information populated with product data is still there.

## Screenshots
| Before            |  After |
:-------------------------:|:-------------------------:
![Before](https://github.com/user-attachments/assets/d845c589-c2e3-432d-b66b-bcee2478ee41) | ![After](https://github.com/user-attachments/assets/9a5fecc5-5e41-4a22-aae7-646aaeb5ad24)

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

pmusolino committed:
remove: unused localized string

pmusolino committed:
fix: Populate the Blaze Campaign tagline and description with product information, and do not fetch AI suggestions again if the user returns to the form view.

pmusolino committed:
update: release-notes 21.5

pmusolino committed:
fix: removed print

pmusolino committed:
update: added PR link to release-notes 21.5

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:
```diff
@@ -323,7 +323,8 @@ final class BlazeCampaignCreationFormViewModel: ObservableObject {
     func onLoad() async {
         await withTaskGroup(of: Void.self) { group in
-            if suggestions.isEmpty {
+            // Make sure AI Suggestions are called just if suggestions are empty, and we didn't already populated tagline, description and ctaText.
+            if suggestions.isEmpty && tagline.isEmpty && description.isEmpty && ctaText.isEmpty {
```
I don't quite understand this part, could you explain why we need the change? From what I understand, `onLoad` is triggered with `task`, i.e. when the creation form appears. There's no other trigger to be prevented.

pmusolino replied:
Now, those fields are populated even if the API call fails. Previously, if the API call failed and you navigated to another screen, like moving forward in the navigation, the `loadAISuggestions` would be called again because `suggestions` was empty. However, we now want to avoid calling AI suggestions again if the ad was pre-populated with product information. This prevents unnecessary UI updates, which can be confusing for the user. If the network request initially fails and we populate it with product info, the user might not notice that the ad information was updated with AI suggestions the second time.

itsmeichigo replied:
Gotcha, so this can be triggered again if tapping Confirm Details and then navigating back 👍

itsmeichigo commented on WooCommerce/Classes/ViewRelated/Blaze/CampaignCreation/BlazeCampaignCreationFormViewModel.swift:
```diff
@@ -355,6 +356,11 @@ final class BlazeCampaignCreationFormViewModel: ObservableObject {
                 ctaText = firstSuggestion.ctaText
             }
         } catch {
+            if let name = product?.name, let shortDescription = product?.shortDescription {
+                tagline = name
+                description = shortDescription
+                ctaText = Localization.shopNow
+            }
```
Should this be the other way around? I have a product without short description and the pre-populated tagline makes it look odd:

<img src="https://github.com/user-attachments/assets/4959b047-ba4d-4298-99a9-6e1b82aee7e8" width=320 />

pmusolino replied:
Do you mean having description = product title and tagline = shortDescription?
From the original issue instructions we have: 
```
Use product name for ad title
Use product short description if available or Description for ad description.
```
then I assume the description it's the tagline here (which appear on top of the "title").

itsmeichigo replied:
Thanks for the explanation. Maybe @JorgeMucientes can confirm the requirements?
Either way, I think we should display the description when short description is not available.

pmusolino replied:
I just checked the "Edit Ad" screen, and I think it should remain this way. In fact, the `tagline` has a limited number of characters, which is fewer than the number of characters allowed in the `description`.
I agree on displaying the description, if short description is not available: 38e93e148e6423131fe252495b2c4247664d15e4

pmusolino committed:
feat: display the description, if short description is not available

itsmeichigo commented:
Another thing: should we hide or update the "Suggested by AI" text and the arrow buttons in the Edit Ad screen when we use the pre-populated content? Should we replace these with the option to retry fetching AI recommendations? 

<img src="https://github.com/user-attachments/assets/cc119e46-e59e-4428-bd33-ca6abc14714d" width=320 />

pmusolino committed:
Merge branch 'trunk' into issue/14616-set-default-values-for-blaze-ad-preview-when-fetching-suggestions-fail

pmusolino committed:
Merge commit '11e5e14c7e76dcd2d5338382a536e8e8f9819fc4' into issue/14616-set-default-values-for-blaze-ad-preview-when-fetching-suggestions-fail

pmusolino committed:
- Update `BlazeEditAdView` to conditionally display AI suggestions based on `isUsingAISuggestions`
- Modify `BlazeEditAdView_Previews` in `BlazeEditAdView` to include `isUsingAISuggestions` parameter
- Add `isUsingAISuggestions` property to `BlazeEditAdViewModel` and initialize it in the constructor
- Pass `isUsingAISuggestions` parameter in `BlazeCampaignCreationFormViewModel`

pmusolino committed:
Update: `BlazeEditAdViewModelTests` to include `isUsingAISuggestions: true` in multiple test cases.

pmusolino committed:
Merge branch 'trunk' into issue/14616-set-default-values-for-blaze-ad-preview-when-fetching-suggestions-fail

dangermattic commented:
<!--
  0 Errors
  0 Warnings
  0 Messages
  1 Markdown
-->

## Danger has errored
[!] Invalid `Dangerfile` file: SIGHUP
```
 #  from Dangerfile:6
 #  -------------------------------------------
 #  # `files: []` forces rubocop to scan all files, not just the ones modified in the PR
 >  rubocop.lint(files: [], force_exclusion: true, inline_comment: true, fail_on_inline_comment: true, include_cop_names: true)
 #  
 #  -------------------------------------------
```

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

pmusolino commented:
Thanks, @itsmeichigo, for raising the point about suggestions by AI in the Edit Screen. I applied the changes in commit 7dd227c46327dcd70b48dd024965aff95c475ba7, and now the "suggested by AI" section will be hidden if the suggestions don't come from AI. Please feel free to take another look.

itsmeichigo reviewed (approved):
Tested again and confirmed that the Edit ad screen is no longer displaying the "Suggested by AI" text 👍 

<img src="https://github.com/user-attachments/assets/59d5c30e-a769-4b35-a9fa-9fe34504f9c7" width=320 />
